sphinx/domains/python.py

```diff
@@ -1195,8 +1237,11 @@ def resolve_xref(self, env: BuildEnvironment, fromdocname: str, builder: Builder
                                 type, searchmode)
 
         if not matches and type == 'attr':
-            # fallback to meth (for property)
+            # fallback to meth (for property; Sphinx-2.4.x)
             matches = self.find_obj(env, modname, clsname, target, 'meth', searchmode)
+        if not matches and type == 'meth':
+            # fallback to attr (for property)
+            matches = self.find_obj(env, modname, clsname, target, 'attr', searchmode)
```

Comment:
As a fallback for `meth` role, python domain searches the property using `attr` role. So it might hit an incorrect one.

Reply:
So as I understand it; the first `if` is for:

* Users who still use `..method` to define their properties
* Users loading an intersphinx file generated by an older sphinx

While the latter `if` is for:

* Users who still use `:meth:` to refer to properties

Does it make sense to add a check that the object being referred to is actually a property?

Reply:
I think a "prop" role is a polite way to do that. But it seems all of you don't like to introduce it. Please don't ask me.

Reply:
Ah, I finally understand what you're getting at. There's no way for sphinx to know if `` :meth:`maybe_prop` `` is a property  or not, as it can't see the old-style `:property:` marker; that information doesn't make it into `objects.inv`.

What I think I'm suggesting is that a warning should be emitted. I'll make another comment with a better set of lines selected...

Reply:
Actually, this is what I'd suggest:
```suggestion
            # this ensures that old versions of sphinx can keep referencing properties
            # from new objects.inv files
            matches = self.find_obj(env, modname, clsname, target, 'attr', searchmode)
            matches = [match for match in matches if match[1] == 'property']
```

Reply:
Why do you reinvent the prop role? Is it better to implement it for internal use?

Reply:
Adding a prop role wouldn't help, because this `if` block is dealing with old rST files which are incorrectly using a `:meth:` role. If you add a new `:prop:` role, these old files will still incorrectly use a `:meth:` role.

Conversely, new rST files have nothing to gain from a `:prop:` role, as the existing `:attr:` role is all they need.

The messiness of the code I suggest is simply due to trying to be backwards compatible. In sphinx 5, you could remove both of these if statements if you wanted.

Reply:
I'd like to say use a prop role as a fallback of the `:meth:` role in the old reST files. It behaves the same as what your suggested code doing.
```
            matches = self.find_obj(env, modname, clsname, target, 'attr', searchmode)
            matches = [match for match in matches if match[1] == 'property']
```
and
```
            matches = self.find_obj(env, modname, clsname, target, 'prop', searchmode)
```
will be equivalent.

Reply:
Am I right in thinking that with that change, users will be able to type `` :prop:`...` `` in their rST files though, which is what we were trying to avoid? Or does "role" mean two different things here, one internal to sphinx and one exposed through rST

If the latter, how about `_prop_backcompat` to make it clear that this is for internal use?

Reply:
Yes, I meant the latter one. I feel `_prop_backcompat` is too much. Adding a comment is enough helping us.

Reply:
How about just `_prop` then with a comment?

Reply:
Okay, go with this.